File: backend/utils/validation.ts
/**
 * Validation Utilities
 * Pure validation functions without framework dependencies
 */

/**
 * Validate email format
 */
export function validateEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Validate password strength
 */
export function validatePassword(password: string): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  if (password.length < 8) {
    errors.push('Password must be at least 8 characters long');
  }

  if (!/[A-Z]/.test(password)) {
    errors.push('Password must contain at least one uppercase letter');
  }

  if (!/[a-z]/.test(password)) {
    errors.push('Password must contain at least one lowercase letter');
  }

  if (!/[0-9]/.test(password)) {
    errors.push('Password must contain at least one number');
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Validate UUID format
 */
export function validateUUID(id: string): boolean {
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return uuidRegex.test(id);
}

/**
 * Sanitize string input
 */
export function sanitizeString(input: string): string {
  // Remove potentially dangerous characters
  return input
    .replace(/<script[^>]*>.*?<\/script>/gi, '')
    .replace(/<[^>]+>/g, '')
    .trim();
}

/**
 * Validate array length
 */
export function validateArrayLength(
  array: any[],
  min?: number,
  max?: number
): boolean {
  if (min !== undefined && array.length < min) {
    return false;
  }

  if (max !== undefined && array.length > max) {
    return false;
  }

  return true;
}

/**
 * Validate number range
 */
export function validateNumberRange(
  value: number,
  min?: number,
  max?: number
): boolean {
  if (min !== undefined && value < min) {
    return false;
  }

  if (max !== undefined && value > max) {
    return false;
  }

  return true;
}

/**
 * Validate enum value
 */
export function validateEnum<T extends string>(
  value: string,
  allowedValues: T[]
): boolean {
  return allowedValues.includes(value as T);
}

/**
 * Validate date format
 */
export function validateDate(dateString: string): Date | null {
  const date = new Date(dateString);

  if (isNaN(date.getTime())) {
    return null;
  }

  return date;
}

/**
 * Validate date range
 */
export function validateDateRange(start: Date, end: Date): boolean {
  return start <= end;
}

File: backend-python/app/utils/validation.py
"""
Validation Utilities
Common validation functions
"""

import re
from uuid import UUID


def validate_email(email: str) -> bool:
    """
    Validate email format.

    Args:
        email: Email address to validate

    Returns:
        True if valid email format
    """
    if not email:
        return False
    email_regex = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
    return bool(re.match(email_regex, email))


def validate_uuid(value: str) -> bool:
    """
    Validate UUID format.

    Args:
        value: String to validate

    Returns:
        True if valid UUID format
    """
    if not value:
        return False
    try:
        UUID(value)
        return True
    except ValueError:
        return False


def validate_password(password: str, min_length: int = 8, require_complexity: bool = True) -> dict[str, any]:
    """
    Validate password strength.

    Args:
        password: Password to validate
        min_length: Minimum password length
        require_complexity: Whether to require uppercase, lowercase, and numbers

    Returns:
        Dict with 'valid' boolean and optional 'message' string
    """
    if not password:
        return {"valid": False, "message": "Password is required"}

    if len(password) < min_length:
        return {"valid": False, "message": f"Password must be at least {min_length} characters long"}

    if require_complexity:
        if not re.search(r"[A-Z]", password):
            return {"valid": False, "message": "Password must contain at least one uppercase letter"}

        if not re.search(r"[a-z]", password):
            return {"valid": False, "message": "Password must contain at least one lowercase letter"}

        if not re.search(r"[0-9]", password):
            return {"valid": False, "message": "Password must contain at least one number"}

    return {"valid": True}


def sanitize_string(value: str | None, max_length: int = 255) -> str | None:
    """
    Sanitize string input.

    Args:
        value: String to sanitize
        max_length: Maximum allowed length

    Returns:
        Sanitized string or None
    """
    if value is None:
        return None
    # Strip whitespace and limit length
    sanitized = value.strip()[:max_length]
    return sanitized if sanitized else None
